### PR TITLE
SPIRV: Decorate matrices and arrays with their strides

### PR DESCRIPTION
**Pull request by jekstrand**

### CODE REVIEW

File: SPIRV/GlslangToSpv.cpp [hunk 1]
```diff
@@ -1459,11 +1465,29 @@ spv::Id TGlslangToSpvTraverser::convertGlslangToSpvType(const glslang::TType& ty
...
     }
 
     return spvType;
 }
 
+// Given an array type, returns the integer stride required for that array
+int TGlslangToSpvTraverser::getArrayStride(const glslang::TType& arrayType)
+{
+    glslang::TType derefType(arrayType, 0);
+    int size;
```
Comment (xorgy) [inline SPIRV/GlslangToSpv.cpp]: It's generally good style to place declarations at the top of the block.

**Comment by johnkslang** [comment]
@jekstrand, looks good, except perhaps missing test result updates.  I'll look at adding those myself.

**Comment by johnkslang** [comment]
Added test changes: 358dd560c516b2f71ae3c855dbd3d3f83294a18c.

File: SPIRV/GlslangToSpv.cpp [hunk 2]
```diff
@@ -1459,11 +1465,29 @@ spv::Id TGlslangToSpvTraverser::convertGlslangToSpvType(const glslang::TType& ty
         } else
             arraySize = type.getOuterArraySize();
         spvType = builder.makeArrayType(spvType, arraySize);
+        builder.addDecoration(spvType, spv::DecorationArrayStride, getArrayStride(type));
```
Comment (johnkslang) [inline SPIRV/GlslangToSpv.cpp]: This code is executed for all arrays.  However, the decoration is only required on transparent (e.g., uniform block) arrays, and indeed not even well-defined for an array of opaque types (e.g., array of samplers).  I'm getting asserts on it, likely due to the latter.  Am looking at fixing.

Comment (johnkslang) [inline SPIRV/GlslangToSpv.cpp]: Fixed with 31ed4830d15d9bd019869b5e04a053a665952f96.

**Pull request merged**